src/workerd/api/node/crypto_dh.c++

```diff
@@ -0,0 +1,255 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+// Copyright Joyent and Node contributors. All rights reserved. MIT license.
+
+#include "crypto.h"
+#include <openssl/bn.h>
+#include <openssl/dh.h>
+#include <workerd/api/crypto-impl.h>
+#include <workerd/jsg/jsg.h>
+
+namespace workerd::api::node {
+
+namespace {
+// Returns a function that can be used to create an instance of a standardized
+// Diffie-Hellman group.
+BIGNUM* (*FindDiffieHellmanGroup(const char* name))(BIGNUM*) {
+#define V(n, p)                                                                \
+  if (strncasecmp(name, n, 7) == 0) {return p;}
+  V("modp14", BN_get_rfc3526_prime_2048);
+  V("modp15", BN_get_rfc3526_prime_3072);
+  V("modp16", BN_get_rfc3526_prime_4096);
+  V("modp17", BN_get_rfc3526_prime_6144);
+  V("modp18", BN_get_rfc3526_prime_8192);
+#undef V
+  return nullptr;
+}
+} // namespace
+
+jsg::Ref<CryptoImpl::DHHandle> CryptoImpl::DHGroupHandle(kj::String name) {
+  return jsg::alloc<CryptoImpl::DHHandle>(name);
+}
+
+jsg::Ref<CryptoImpl::DHHandle> CryptoImpl::DHHandle::constructor(
+    jsg::Lock &js, kj::OneOf<kj::Array<kj::byte>, int> sizeOrKey,
+    kj::OneOf<kj::Array<kj::byte>, int> generator) {
+  return jsg::alloc<CryptoImpl::DHHandle>(sizeOrKey, generator);
+}
+
+bool CryptoImpl::DHHandle::VerifyContext() {
+  int codes;
+  if (!DH_check(dh, &codes))
+    return false;
+  verifyError = codes;
+  return true;
+}
+
+CryptoImpl::DHHandle::DHHandle(kj::OneOf<kj::Array<kj::byte>, int> &sizeOrKey,
+                               kj::OneOf<kj::Array<kj::byte>, int> &generator)
+                               : dh(nullptr), verifyError(0) {
+  JSG_REQUIRE(Init(sizeOrKey, generator), Error, "failed to init DH");
+}
+
+CryptoImpl::DHHandle::DHHandle(kj::String& name) : dh(nullptr), verifyError(0) {
+  JSG_REQUIRE(InitGroup(name), Error, "failed to init DH");
+}
+
+bool CryptoImpl::DHHandle::InitGroup(kj::String& name) {
+  auto group = FindDiffieHellmanGroup(name.begin());
+  JSG_REQUIRE(group != nullptr, Error, "Failed to init DHGroup: invalid group");
+  auto groupKey = group(nullptr);
+  KJ_ASSERT(groupKey != nullptr);
+
+  const int kStandardizedGenerator = 2;
+  dh = DH_new();
+
+  BIGNUM* bn_g = BN_new();
```

Comment:
To avoid the mistake I made earlier, let's add a comment in these mentioning that `DH_set0_pqg` takes ownership of the pointer if the call was successful to explain why `kj::Own` isn't being used.